packages/contracts/src/contracts/current/forwarder/MixinForwarderExpectedResults.sol

```diff
@@ -0,0 +1,114 @@
+pragma solidity ^0.4.24;
+pragma experimental ABIEncoderV2;
+
+import "./MixinForwarderCore.sol";
+
+contract MixinForwarderExpectedResults is MixinForwarderCore {
+
+    /// @dev Simulates the 0x Exchange fillOrder validation and calculations, without performing any state changes.
+    /// @param order An Order struct containing order specifications.
+    /// @param takerAssetFillAmount A number representing the amount of this order to fill.
+    /// @return fillResults Amounts filled and fees paid by maker and taker.
+    function calculateFillOrderFillResults(Order memory order, uint256 takerAssetFillAmount)
+        public
+        view
+        returns (Exchange.FillResults memory fillResults)
+    {
+        Exchange.OrderInfo memory orderInfo = EXCHANGE.getOrderInfo(order);
+        uint8 _status;
+        (_status, fillResults) = EXCHANGE.calculateFillResults(
+            order,
+            orderInfo.orderStatus,
+            orderInfo.orderTakerAssetFilledAmount,
+            takerAssetFillAmount
+        );
+        return fillResults;
+    }
+
+    /// @dev Calculates a FillResults total for selling takerAssetFillAmount over all orders. 
+    ///      Including the fees required to be paid. 
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param takerAssetFillAmount A number representing the amount of this order to fill.
+    /// @return totalFillResults Amounts filled and fees paid by maker and taker.
+    function calculateMarketSellFillResults(Order[] memory orders, uint256 takerAssetFillAmount)
+        internal
+        view
+        returns (Exchange.FillResults memory totalFillResults)
+    {
+        for (uint256 i = 0; i < orders.length; i++) {
+            require(areBytesEqual(orders[i].makerAssetData, orders[0].makerAssetData), SAME_ASSET_TYPE_REQUIRED);
+            uint256 remainingTakerAssetFillAmount = safeSub(takerAssetFillAmount, totalFillResults.takerAssetFilledAmount);
+
+            Exchange.FillResults memory fillOrderExpectedResults = calculateFillOrderFillResults(orders[i], remainingTakerAssetFillAmount);
+
+            addFillResults(totalFillResults, fillOrderExpectedResults);
+            if (totalFillResults.takerAssetFilledAmount == takerAssetFillAmount) {
```

Comment:
Is it possible to have a small margin of error?

Reply:
Can you go into more detail about the small margin of error you potentially see?